[BE] Make maybe_aliasing_or_mutating proper tag

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #131990

For better tracking, we need to make maybe aliasing/mutating ops with proper tag. We need to special case native_batch_norm because it is not a CIA but has a wrong schema. I guess native_batch_norm will be removed at some point, so until then we just keep it around. 

D60347117

tugsbayasgalan committed:
[BE] Make maybe_aliasing_or_mutating proper tag

For better tracking, we need to make maybe aliasing/mutating ops with proper tag.

Differential Revision: [D60347117](https://our.internmc.facebook.com/intern/diff/D60347117/)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -4307,6 +4315,7 @@
     CUDA: batch_norm_cuda
     MPS: batch_norm_mps
     MkldnnCPU: mkldnn_batch_norm
+  tags: maybe_aliasing_or_mutating
```
I'm not sure about this, in these cases the schema is just wrong (as opposed to composites, where schema is not meaningful)

tugsbayasgalan replied:
Do you mean we don't need this tag at all or we need two seperate tags (one for CIA ops that are maybe mutating and one for primitive ops)?

bdhirsh commented on aten/src/ATen/native/tags.yaml:
```diff
@@ -63,3 +63,8 @@
+  desc: |
+          For some ops, we can't statically determine whether the op is functional or not. It is useful to
+          know this information for export as we would want to decompose these ops as they are unsafe to be
+          preserved.
```
Maybe worth a comment that this tag should only apply to ops that decompose before functionalization/autograd (CompositeImplicitAutograd). Since autograd/functionalization need to statically know schema information.

tugsbayasgalan commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -1859,6 +1865,7 @@
   dispatch:
     CUDA: cudnn_batch_norm
   autogen: cudnn_batch_norm.out
+  tags: maybe_aliasing_or_mutating
```
@bdhirsh  This is non-CIA op with wrong schema. After @andrewor14 's work on consolidation, can we expect to never see this in PT2? In that case, i think it is ok to just not tag them and only tag CIA ones

tugsbayasgalan committed:
Update on "[BE] Make maybe_aliasing_or_mutating proper tag"


For better tracking, we need to make maybe aliasing/mutating ops with proper tag.

Differential Revision: [D60347117](https://our.internmc.facebook.com/intern/diff/D60347117/)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[BE] Make maybe_aliasing_or_mutating proper tag"


For better tracking, we need to make maybe aliasing/mutating ops with proper tag.

Differential Revision: [D60347117](https://our.internmc.facebook.com/intern/diff/D60347117/)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60347117).

tugsbayasgalan committed:
Update on "[BE] Make maybe_aliasing_or_mutating proper tag"


For better tracking, we need to make maybe aliasing/mutating ops with proper tag. We need to special case native_batch_norm because it is not a CIA but has a wrong schema. I guess native_batch_norm will be removed at some point, so until then we just keep it around. 

Differential Revision: [D60347117](https://our.internmc.facebook.com/intern/diff/D60347117/)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60347117).

tugsbayasgalan committed:
Update on "[BE] Make maybe_aliasing_or_mutating proper tag"


For better tracking, we need to make maybe aliasing/mutating ops with proper tag. We need to special case native_batch_norm because it is not a CIA but has a wrong schema. I guess native_batch_norm will be removed at some point, so until then we just keep it around. 

Differential Revision: [D60347117](https://our.internmc.facebook.com/intern/diff/D60347117/)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60347117).

bdhirsh commented on torch/_decomp/__init__.py:
```diff
@@ -48,6 +48,24 @@
+    """
+    if not isinstance(op, torch._ops.OpOverload):
+        return False
+    if torch.Tag.maybe_aliasing_or_mutating in op.tags:
```
Can we just give this tag to `native_batch_norm` directly so it doesn't need to be special-cased here?

tugsbayasgalan replied:
I think the problem here is native_batch_norm is not a CIA op so basically its schema is just wrong.

bdhirsh reviewed (commented):
looks like there are still some failing tests?

tugsbayasgalan commented:
> looks like there are still some failing tests?

Yup i think most of them due to bad trunk tho

tugsbayasgalan committed:
Update

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[BE] Make maybe_aliasing_or_mutating proper tag"


For better tracking, we need to make maybe aliasing/mutating ops with proper tag. We need to special case native_batch_norm because it is not a CIA but has a wrong schema. I guess native_batch_norm will be removed at some point, so until then we just keep it around. 

Differential Revision: [D60347117](https://our.internmc.facebook.com/intern/diff/D60347117/)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60347117).

tugsbayasgalan commented:
@pytorchbot rebase

tugsbayasgalan committed:
Update

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[BE] Make maybe_aliasing_or_mutating proper tag"


For better tracking, we need to make maybe aliasing/mutating ops with proper tag. We need to special case native_batch_norm because it is not a CIA but has a wrong schema. I guess native_batch_norm will be removed at some point, so until then we just keep it around. 

Differential Revision: [D60347117](https://our.internmc.facebook.com/intern/diff/D60347117/)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D60347117).

bdhirsh commented on torch/_export/__init__.py:
```diff
@@ -169,9 +170,23 @@ def capture_pre_autograd_graph(
+
+        # We force create native_batch_norm because the below materialization logic
+        # only applies to CIA ops.
+        maybe_aliasing_or_mutating_ops = [torch.ops.aten.native_batch_norm.default]
```
it looks like you're hardcoding `aten.native_batch_norm` in two places: both here, and in your helper `_should_decompose_because_unsafe_op(op)`. Can't you just re-use that `_should_decompose_because_unsafe_op()` helper in ordre to generate this list?

bdhirsh reviewed (approved):
left some comments!

izaitsevfb commented:
@pytorchbot merge

izaitsevfb commented:
@pytorchbot merge

izaitsevfb commented:
@pytorchbot merge -i
